server/src/main/java/org/elasticsearch/common/geo/TriangleTreeReader.java

```diff
@@ -0,0 +1,393 @@
+            rectangle2D.setValues(minX, maxX, minY, maxY);
+            byte metadata = input.readByte();
+            if ((metadata & 1 << 2) == 1 << 2) {
+                int x = Math.toIntExact(thisMaxX - input.readVLong());
+                int y = Math.toIntExact(thisMaxY - input.readVLong());
+                if (rectangle2D.contains(x, y)) {
+                    return GeoRelation.QUERY_CROSSES;
+                }
+                thisMinX = x;
+            } else if ((metadata & 1 << 3) == 1 << 3) {
+                int aX = Math.toIntExact(thisMaxX - input.readVLong());
+                int aY = Math.toIntExact(thisMaxY - input.readVLong());
+                int bX = Math.toIntExact(thisMaxX - input.readVLong());
+                int bY = Math.toIntExact(thisMaxY - input.readVLong());
+                if (rectangle2D.intersectsLine(aX, aY, bX, bY)) {
+                    return GeoRelation.QUERY_CROSSES;
+                }
+                thisMinX = aX;
+            } else {
+                int aX = Math.toIntExact(thisMaxX - input.readVLong());
+                int aY = Math.toIntExact(thisMaxY - input.readVLong());
+                int bX = Math.toIntExact(thisMaxX - input.readVLong());
+                int bY = Math.toIntExact(thisMaxY - input.readVLong());
+                int cX = Math.toIntExact(thisMaxX - input.readVLong());
+                int cY = Math.toIntExact(thisMaxY - input.readVLong());
+                boolean ab = (metadata & 1 << 4) == 1 << 4;
+                boolean bc = (metadata & 1 << 5) == 1 << 5;
+                boolean ca = (metadata & 1 << 6) == 1 << 6;
+                rel = rectangle2D.relateTriangle(aX, aY, ab, bX, bY, bc, cX, cY, ca);
+                if (rel == GeoRelation.QUERY_CROSSES) {
+                    return GeoRelation.QUERY_CROSSES;
+                }
+                thisMinX = aX;
+            }
+            if ((metadata & 1 << 0) == 1 << 0) { // left != null
+                GeoRelation left = relate(rectangle2D, false, thisMaxX, thisMaxY);
+                if (left == GeoRelation.QUERY_CROSSES) {
+                    return GeoRelation.QUERY_CROSSES;
+                } else if (left == GeoRelation.QUERY_INSIDE) {
+                    rel = left;
+                }
+            }
+            if ((metadata & 1 << 1) == 1 << 1) { // right != null
+                if (rectangle2D.maxX >= thisMinX) {
+                    GeoRelation right = relate(rectangle2D, false, thisMaxX, thisMaxY);
+                    if (right == GeoRelation.QUERY_CROSSES) {
+                        return GeoRelation.QUERY_CROSSES;
+                    } else if (right == GeoRelation.QUERY_INSIDE) {
+                        rel = right;
+                    }
+                }
+            }
+        }
+        return rel;
+    }
+
+    private GeoRelation relate(Rectangle2D rectangle2D, boolean splitX, int parentMaxX, int parentMaxY) throws IOException {
```

Comment:
I know there could be more documentation in the EdgeTreeReader's relate, but I think it would be nice to add some in-line comments explaining what is being checked in each branch here